### PR TITLE
Port over bugsnag dot net classes

### PR DESCRIPTION
**Pull request by mglodack**
#### What's in here?

Bugsnag portable class library.

##### Differences from `Bugsnag.NET`
- Portable class library does not support grabbing device specific information
- Device information has been removed from `Bugsnag.PCL`

##### Solution to posting device information
- Create a custom tab in bugsang using the metaData property on the json object as seen [here](https://bugsnag.com/docs/notifier-api#json-payload)

**NOTE:** An example will be posted on how to pass in metaData.

### CODE REVIEW

File: Bugsnag.UWP/Bugsnag.UWP.csproj [hunk 1]
```diff
@@ -107,9 +107,32 @@
...
+    <Compile Include="Request\Interfaces\IUser.cs" />
+    <Compile Include="Request\Notice.cs" />
+    <Compile Include="Request\Notifier.cs" />
+    <Compile Include="Request\Severity.cs" />
+    <Compile Include="Request\StackTraceLine.cs" />
+    <Compile Include="Request\Thread.cs" />
+    <Compile Include="Request\User.cs" />
     <EmbeddedResource Include="Properties\Bugsnag.UWP.rd.xml" />
   </ItemGroup>
+  <ItemGroup />
```
Comment (awseward) [inline Bugsnag.UWP/Bugsnag.UWP.csproj]: Can we remove this?

File: Bugsnag.UWP/Bugsnag.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        public static string ApiKey { get; set; }
+
+        static INotifier _notifier = new Notifier();
+
+        public static INotifier Notifier
+        {
+            get { return _notifier; }
+            set { _notifier = value; }
+        }
```
Comment (awseward) [inline Bugsnag.UWP/Bugsnag.cs]: Would this kind of syntax work here for these?

```cs
public static INotifier Notifier { get; set; } = new Notifier();
```

File: Bugsnag.UWP/Request/Device.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        string _osVersion = new Func<string>(() =>
+        {
+            var  sv = AnalyticsInfo.VersionInfo.DeviceFamilyVersion;
+            ulong v = ulong.Parse(sv);
+            ulong v1 = (v & 0xFFFF000000000000L) >> 48;
+            ulong v2 = (v & 0x0000FFFF00000000L) >> 32;
+            ulong v3 = (v & 0x00000000FFFF0000L) >> 16;
+            ulong v4 = v & 0x000000000000FFFFL;
+            return $"{v1}.{v2}.{v3}.{v4}";
+        }).Invoke();
```
Comment (awseward) [inline Bugsnag.UWP/Request/Device.cs]: Well this is lovely...

File: Bugsnag.UWP/Request/Event.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        public Event(IEnumerable<Exception> unwrapped)
+        {
+            Errors = _GetErrors(unwrapped);
+            GroupingHash = _GetGroupingHash(unwrapped);
+            Context = _GetContext(unwrapped);
+        }
+
+        static string _uwpUnknownMethod = "UWP_UNKNOWN_METHOD";
```
Comment (awseward) [inline Bugsnag.UWP/Request/Event.cs]: Might as well make this `readonly`, yeah?

File: Bugsnag.UWP/Request/Interfaces/IEvent.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        string Context { get; }
+
+        [JsonProperty("groupingHash")]
+        string GroupingHash { get; }
+
+        [JsonProperty("severity")]
+        string Severity { get; }
+
+        [JsonProperty("user")]
+        IUser User { get;}
```
Comment (awseward) [inline Bugsnag.UWP/Request/Interfaces/IEvent.cs]: Space, please.

File: Bugsnag.UWP/Request/Thread.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+        public string Name
+        {
+            get { throw new NotImplementedException(); }
+        }
+
+        public IEnumerable<IStackTraceLine> Stacktrace
+        {
+            get { throw new NotImplementedException(); }
+        }
```
Comment (awseward) [inline Bugsnag.UWP/Request/Thread.cs]: I'm assuming these will be implemented eventually...

**Comment by awseward** [comment]
I've noticed that the `Newtonsoft.Json` referenced in here is 8.x, where I believe the existing project depended on 6.x...

It might be nice to have their dependencies match, but it also might not even matter. Curious to hear your thoughts on this.

**Comment by mglodack** [comment]
@awseward With regards to `Newtonsoft.Json`. 

I would prefer to be on the latest for this project, and I will bump the `Bugsnag.NET` project with a separate pull request.

File: Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        [JsonProperty("columnNumber")]
+        int? ColumnNumber { get; }
+
+        [JsonProperty("method")]
+        string Method { get; }
+
+        [JsonProperty("inProject")]
+        bool InProject { get; }
+
+        // TODO: Code
```
Comment (awseward) [inline Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs]: ?

File: Bugsnag.PCL/Request/Notifier.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            var assembly = typeof(Notifier).GetTypeInfo().Assembly;
+            var assemblyName = new AssemblyName(assembly.FullName);
+            return assemblyName.Version.ToString();
+        }).Invoke();
+
+        public string Version { get { return _version; } }
+
+
+        string _url = "https://github.com/awseward/Bugsnag.NET";
```
Comment (awseward) [inline Bugsnag.PCL/Request/Notifier.cs]: Read-only?

**Comment by awseward** [comment]
If we want to allow the consuming app to pass in device info, we should still send it to bugsnag the same way. A custom tab isn't the solution.

File: Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs (hunk 7 above)
Comment (mglodack) [inline Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs]: This was copied over. 

Here is what this should allow in the future. Not in this PR.
```
// The code in this file surrounding this line. You can include
// up to three lines on either side of the line that crashed.
// These will be displayed on the bugsnag dashboard when you expand
// that line.
// (optional)
  code: {
       1231: "  def a",
       1232: "",
       1233: "    if problem?",
       1234: "      raise 'something went wrong'",
       1235: "    end"
       1236: "",
       1237: "  end"
   }
```

File: Bugsnag.PCL/BugsnagSender.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            var json = JsonConvert.SerializeObject(notice, _settings);
+
+            var content = new StringContent(json);
+
+            return await new HttpClient().PostAsync(uri, content);
+        }
+
+        static Uri _GetUri(bool useSSL)
+        {
+            return (useSSL) ? _sslUri : _uri;
```
Comment (awseward) [inline Bugsnag.PCL/BugsnagSender.cs]: Not blocking, but these parens are not necessary.

File: Bugsnag.PCL/Extensions/Extensions.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            return ex.InnerException.Unwrap().Concat(new Exception[] { ex });
+        }
+
+        public static IEnumerable<string> ToLines(this Exception ex)
+        {
+            if (ex == null || ex.StackTrace == null)
+            {
+                return new string[] { String.Empty };
```
Comment (awseward) [inline Bugsnag.PCL/Extensions/Extensions.cs]: NB: Would it make more sense for this to be `Enumerable.Empty<string>();`

File: Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs (hunk 7 above)
Comment (awseward) [inline Bugsnag.PCL/Request/Interfaces/IStackTraceLine.cs]: Sounds good.

**Comment by awseward** [comment]
+1

**Comment by awseward** [comment]
Very happy about that 14k lines of json gone!

File: Bugsnag.PCL/Extensions/Extensions.cs (hunk 10 above)
Comment (IgorBurakov) [inline Bugsnag.PCL/Extensions/Extensions.cs]: It is possible to make the replacement, but need to check if Bugsnag will register that event. I think there should be at least one line to make it work properly

**Pull request merged**